Orange/projection/pca.py

```diff
@@ -201,7 +201,7 @@ def _fit_truncated(self, X, n_components, svd_solver):
                 random_state=random_state,
             )
 
-        self.n_samples_, self.n_features_ = n_samples, n_features
+        self.n_samples_ = n_samples
```

Comment:
I think this works because:

Most of the code in `ImprovedPCA` is copied directly from `sklearn.decomposition.PCA`. This was done because it wasn't possible to add the randomized algorithm that enables PCA on sparse data without modifying these functions.
Until recently, sklearn's PCA used `self.n_features_`. However, they recently switched over to using `self.n_features_in`, which is more generally used in sklearn, and is part of their `BaseEstimator` class. The `self.n_features_in` attribute is set in [`BaseEstimator._check_n_features`](https://github.com/scikit-learn/scikit-learn/blob/0266481aeb736e5f2c9bf07c4ebb016845d50eb9/sklearn/base.py#L318-L362), which is, in turn, called by [`BaseEstimator._validate_data`](https://github.com/scikit-learn/scikit-learn/blob/0266481aeb736e5f2c9bf07c4ebb016845d50eb9/sklearn/base.py#L453-L560). In scikit-learn's PCA, this method is called in [`PCA._fit`](https://github.com/scikit-learn/scikit-learn/blob/0266481aeb736e5f2c9bf07c4ebb016845d50eb9/sklearn/decomposition/_pca.py#L485-L487), which ensures that the `self.n_features_in` attribute is set on sklearn's implementation.

They also deprecated `self.n_features` and replaced it with a [readonly property](https://github.com/scikit-learn/scikit-learn/blob/0266481aeb736e5f2c9bf07c4ebb016845d50eb9/sklearn/decomposition/_pca.py#L408-L414) which just returns `self.n_featuers_in` anyways. This is still the same functionality as before, but you have to trace it through these different methods. So, this is still backwards compatible.

So, we can do the same thing in our `ImprovedPCA` class. We can get rid of `self.n_features_ = features`, since this will be set when we call `self._validate_data`, which is inherited from `sklearn.decomposition.PCA <- sklearn.BaseEstimator`. So, the other change we need is to replace our previous call to [`check_array`](https://github.com/biolab/orange3/blob/master/Orange/projection/pca.py#L102-L108) to `self._validate_data`. And I believe this should ensure the exact same functionality as before.

Reply:
The `transform` method should also probably be updated to reflect how [sklearn is doing things now](https://github.com/scikit-learn/scikit-learn/blob/0266481aeb736e5f2c9bf07c4ebb016845d50eb9/sklearn/decomposition/_base.py#L120).

So, this would mean changing

```python
        X = check_array(
            X,
            accept_sparse=["csr", "csc"],
            dtype=[np.float64, np.float32],
            ensure_2d=True,
            copy=self.copy,
        )
```

on lines 224-230 to

```python
X = self._validate_data(
    X, dtype=[np.float64, np.float32], reset=False, check_params=dict(accept_sparse=["csr", "csc"])
)
```

Reply:
@pavlin-policar, thanks for the explanation.